hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DualFileWriter.java

```diff
@@ -0,0 +1,255 @@
+      if (deleteFamily == null) {
+        deleteFamilyVersionList.add(cell);
+        addLiveVersion(cell);
+      } else {
+        addHistoricalVersion(cell);
+      }
+    } else if (cell.getType() == Cell.Type.DeleteColumn) {
+      if (!isDeletedByDeleteFamily(cell) && deleteColumn == null) {
+        deleteColumn = cell;
+        addLiveVersion(cell);
+      } else {
+        addHistoricalVersion(cell);
+      }
+    } else if (cell.getType() == Cell.Type.Delete) {
+      if (!isDeletedByDeleteFamily(cell) && deleteColumn == null) {
+        deleteColumnVersionList.add(cell);
+        addLiveVersion(cell);
+      } else {
+        addHistoricalVersion(cell);
+      }
+    } else if (cell.getType() == Cell.Type.Put) {
+      if (livePutCellCount < maxVersions) {
+        // This is a live put cell (i.e., the latest version) of a column. Is it deleted?
+        if (!isDeleted(cell)) {
+          addLiveVersion(cell);
+          livePutCellCount++;
+        } else {
+          // It is deleted
+          addHistoricalVersion(cell);
+        }
+      } else {
+        // It is an older put cell
+        addHistoricalVersion(cell);
+      }
+    }
+    lastCell = cell;
+  }
+
+  @Override
+  public void appendAll(List<Cell> cellList) throws IOException {
+    if (!dualWriterEnabled) {
+      // If the dual writer is not enabled then all cells are written to one file. We use
+      // the live version file in this case
+      for (Cell cell : cellList) {
+        addLiveVersion(cell);
+      }
+      return;
+    }
+    if (cellList.isEmpty()) {
+      return;
+    }
+    if (lastCell != null && comparator.compareRows(lastCell, cellList.get(0)) != 0) {
+      // It is a new row and thus time to reset the state
+      initRowState();
+    }
+    for (Cell cell : cellList) {
+      appendCell(cell);
+    }
+  }
+
+  @Override
+  public void append(Cell cell) throws IOException {
+    if (!dualWriterEnabled) {
+      // If the dual writer is not enabled then all cells are written to one file. We use
+      // the live version file in this case
+      addLiveVersion(cell);
+      return;
+    }
+    if (lastCell != null && comparator.compareRows(lastCell, cell) != 0) {
+      // It is a new row and thus time to reset the state
+      initRowState();
+    }
+    appendCell(cell);
+  }
+
+  @Override
+  protected Collection<StoreFileWriter> writers() {
+    return writers;
+  }
+
+  @Override
+  protected void preCommitWriters() throws IOException {
+    if (writers.isEmpty()) {
+      liveVersionWriter = writerFactory.createWriter();
+      writers.add(liveVersionWriter);
+    }
+    if (!dualWriterEnabled) {
+      return;
+    }
+    if (liveVersionWriter != null) {
+      liveVersionWriter.appendFileInfo(HAS_LIVE_VERSIONS_KEY, Bytes.toBytes(true));
+    }
+    if (historicalVersionWriter != null) {
+      historicalVersionWriter.appendFileInfo(HAS_LIVE_VERSIONS_KEY, Bytes.toBytes(false));
+    }
+  }
+
+  public HFile.Writer getHFileWriter() {
```

Comment:
nit: instead of `getHFileWriter()`, maybe we can call it `getLiveVersionHFileWriter()`?